File: app/src/main/java/com/jonlatane/beatpad/harmony/chord/Chord.java
package com.jonlatane.beatpad.harmony.chord;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by jonlatane on 5/5/17.
 */

public class Chord implements Parcelable {
    private static final String TAG = Chord.class.getSimpleName();
    static final String[] NAMES = {"C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"};
    public static final int[] MAJ = {0, 4, 7};
    public static final int[] MAJ_6 = {0, 4, 7, 9};
    public static final int[] MAJ_6_9 = {0, 2, 4, 7, 9};
    public static final int[] MAJ_7 = {0, 4, 7, 11};
    public static final int[] MIN = {0, 3, 7};
    public static final int[] MIN_7 = {0, 3, 7, 10};
    public static final int[] MIN_MAJ_7 = {0, 3, 7, 11};
    public static final int[] DOM_7 = {0, 4, 7, 10};
    public static final int[] DIM = {0, 3, 6};
    public static final int[] AUG = {0, 4, 8};

    public final int root;
    public final int[] extension;
    public final Heptatonics heptatonics;

    public Chord(int root, int[] extension) {
        this.root = (1200 + root) % 12;
        this.heptatonics = new Heptatonics(root);
        this.extension = extension;
        init();
    }

    private Chord(Parcel in) {
        root = in.readInt();
        heptatonics = new Heptatonics(root);
        extension = new int[in.readInt()];
        in.readIntArray(extension);
        init();
    }

    private void init() {
        for(int i = 0; i < extension.length; i++) {
            int tone = (1200 + extension[i]) % 12;
            extension[i] = tone;
            heptatonics.colors.add(tone);
        }
        Arrays.sort(extension);
    }

    public Chord plus(int... newTones) {
        int[] newExtension = new int[extension.length + newTones.length];
        System.arraycopy(extension, 0, newExtension, 0, extension.length);
        System.arraycopy(newTones, 0, newExtension, extension.length, newTones.length);
        return new Chord(root, newExtension);
    }

    /**
     * Gets a derivative chord created by replacing all of the outTones with inTones.
     * For instance, replaceOrAdd(2,1) should turn any chord except a #9 chord into that chord with a b9
     * @param outTone
     * @param inTone
     * @return A new chord.  Its extension will have all of this chord's instances of outTone replaced
     *         with inTone.  If none were found, inTone will be added.
     */
    public Chord replaceOrAdd(int outTone, int inTone) {
        boolean found = false;
        int[] newExtension = new int[extension.length];
        for(int i = 0; i < extension.length; i++) {
            int tone = extension[i];
            if(tone == outTone) {
                found = true;
                newExtension[i] = inTone;
            } else {
                newExtension[i] = tone;
            }
        }
        Chord result = new Chord(root, newExtension);
        if(!found) {
            result = result.plus(inTone);
        }
        return result;
    }

    /**
     *
     * @param bottom lowest allowed note, inclusive
     * @param top highest allowed note, inclusive
     * @return
     */
    public List<Integer> getTones(int bottom, int top) {
        List<Integer> tones = new LinkedList<>();
        int currentRoot = root - 144;
        while(currentRoot + 12 < bottom) {
            currentRoot += 12;
        }
        while(currentRoot + 12 <= top) {
            for(int color : extension) {
                int tone = currentRoot + color;
                if(tone >= bottom && tone <= top) {
                    tones.add(tone);
                }
            }
            currentRoot += 12;
        }
        return tones;
    }

    public String getName() {
        return NAMES[root]+ heptatonics.getColorString();
    }

    public boolean isMinor() {
        return heptatonics.isMinor();
    }

    public boolean isMajor() {
        return heptatonics.isMajor();
    }

    public boolean isDominant() {
        return heptatonics.isDominant();
    }

    public boolean containsTone(int tone) {
        return containsColor(tone - root);
    }

    public boolean containsColor(int color) {
        color = (1200 + color) % 12;
        for(int i : extension) {
            if(color == i) return true;
        }
        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(root);
        dest.writeInt(extension.length);
        dest.writeIntArray(extension);
    }

    public static final Creator<Chord> CREATOR = new Creator<Chord>() {
        @Override
        public Chord createFromParcel(Parcel in) {
            return new Chord(in);
        }

        @Override
        public Chord[] newArray(int size) {
            return new Chord[size];
        }
    };
}

File: app/src/main/kotlin/com/jonlatane/beatpad/harmony/chord/Chord.kt
package com.jonlatane.beatpad.harmony.chord

import android.os.Parcel
import android.os.Parcelable
import com.jonlatane.beatpad.harmony.chord.heptatonics.Heptatonics
import com.jonlatane.beatpad.util.mod12
import com.jonlatane.beatpad.view.melody.BaseMelodyView

class Chord : Parcelable {
	val root: Int
	val extension: IntArray
	val heptatonics: Heptatonics

	constructor(root: Int, extension: IntArray) {
		this.root = root.mod12
		this.extension = extension.map { it.mod12 }.toIntArray()
		this.heptatonics = Heptatonics(setOf(*this.extension.toTypedArray()))
	}

	private constructor(parcel: Parcel) {
		root = parcel.readInt().mod12
		extension = IntArray(parcel.readInt())
		parcel.readIntArray(extension)
		heptatonics = Heptatonics(setOf(*extension.toTypedArray()))
	}

	fun plus(vararg newTones: Int): Chord {
		val newExtension = IntArray(extension.size + newTones.size)
		System.arraycopy(extension, 0, newExtension, 0, extension.size)
		System.arraycopy(newTones, 0, newExtension, extension.size, newTones.size)
		return Chord(root, newExtension)
	}

	/**
	 * Gets a derivative chord created by replacing all of the outTones with inTones.
	 * For instance, replaceOrAdd(2,1) should turn any chord except a #9 chord into that chord with a b9
	 * @param outTone
	 * *
	 * @param inTone
	 * *
	 * @return A new chord.  Its extension will have all of this chord's instances of outTone replaced
	 * *         with inTone.  If none were found, inTone will be added.
	 */
	fun replaceOrAdd(outTone: Int, inTone: Int): Chord {
		var found = false
		val newExtension = IntArray(extension.size)
		for (i in extension.indices) {
			val tone = extension[i]
			if (tone == outTone) {
				found = true
				newExtension[i] = inTone
			} else {
				newExtension[i] = tone
			}
		}
		var result = Chord(root, newExtension)
		if (!found) {
			result = result.plus(inTone)
		}
		return result
	}

	/**
	 * @param bottom lowest allowed note, inclusive
	 * @param top highest allowed note, inclusive
	 */
	fun getTones(bottom: Int = BaseMelodyView.BOTTOM, top: Int = BaseMelodyView.TOP): List<Int> {
		return (bottom..top).filter {
			it.mod12 in extension.map { (root + it).mod12 }
		}
	}

	/**
	 * Retrieves the closest tone in this chord to the given tone.
	 * If two tones in the chord are equally close, returns the lower one.
	 */
	fun closestTone(tone: Int, bottom: Int = BaseMelodyView.BOTTOM, top: Int = BaseMelodyView.TOP): Int {
		return getTones(Math.max(bottom, tone - 12), Math.min(top, tone + 12)).minBy {
			Math.abs(tone - it)
		}!!
	}

	val name: String
		get() = mod12Names[root] + heptatonics.colorString

	fun containsTone(tone: Int): Boolean {
		return containsColor(tone - root)
	}

	fun containsColor(color: Int): Boolean {
		return extension.contains(color.mod12)
	}

	val isMinor: Boolean get() = heptatonics.isMinor
	val isMajor: Boolean get() = heptatonics.isMajor
	val isAugmented: Boolean get() = isMajor && heptatonics.fifth == AUGMENTED
	val isDiminished: Boolean get() = isMinor && heptatonics.fifth == DIMINISHED
	val isSus: Boolean get() = heptatonics.isSus
	val hasMinor7: Boolean get() = heptatonics.hasMinor7
	val hasMajor7: Boolean get() = heptatonics.hasMajor7
	val hasDiminished5: Boolean get() = heptatonics.hasDiminished5
	val hasAugmented5: Boolean get() = heptatonics.hasAugmented5
	val isDominant: Boolean get() = heptatonics.isDominant

	companion object {
		private val mod12Names = arrayOf("C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B")
		private val mod7Names = arrayOf("C", "D", "E", "F", "G", "A", "B")

		@JvmField
		val CREATOR: Parcelable.Creator<Chord> = object : Parcelable.Creator<Chord> {
			override fun createFromParcel(parcel: Parcel): Chord {
				return Chord(parcel)
			}

			override fun newArray(size: Int): Array<Chord?> {
				return arrayOfNulls(size)
			}
		}
	}

	override fun describeContents(): Int {
		return 0
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		dest.writeInt(root)
		dest.writeInt(extension.size)
		dest.writeIntArray(extension)
	}
}
